protocols/request-response/src/lib.rs

```diff
@@ -696,6 +723,17 @@ where
                 }
             }
         }
+        // Its possible that this dial failure is for an existing connection.
```

Comment:
This comment is somewhat unnecessary because there is no such thing as a "non-existing" connection.

If we wanted to be precise we could explain that, depending on where this behaviour is in the tree, we could have already altered our state for this ID but another behaviour "denied" the connection after us.

Now that I am writing this, I had the idea that we can also fix this by only alterting the state in the `ConnectionEstablished` event.

Perhaps that might be the better solution after all?

Reply:
Perhaps we should extend the docs of `NetworkBehaviour` to explain that.

cc @mxinden This can actually a bit of a problem that didn't occur to me before. Our strategy of "preloading" handlers can lead to us losing state if a `NetworkBehaviour` "after" us denies a connection.

To be safe, these "connection management" plugins should always come first in the behaviour tree. But also, we currently don't call `poll_close` on the handler when the connection is denied which means a handler doesn't have a way of giving the events back to the behaviour. But that is not an elegant solution anyway because it would mean _each_ protocol has to deal with thes lifecycle bugs.

I have to think about this some more but the better solution might be to change the `handle_established` callbacks to take `&self` and dispatch the `ConnectionEstablished` event only _after_ the handlers have been fully created.

That would avoid this partial altering of state at compile-time.

Reply:
I don't know enough of the details of how preloading handlers works, however it was very confusing issue to debug that a connection that was "established" could have a listen or dial failure. I had wrongly assumed that once established a connection could fail but that would not be communicated as a listen/dial failure but rather just a connection error/closed message. 

So anything that makes that relationship clear would be good.

Reply:
Yeah it didn't realise this issue until I thought deeper about this PR :(

It is unfortunately somewhat of a design flaw. Because `NetworkBehaviour`s are composed into a tree AND the `handle_` functions take `&mut self`, you might end up modifying our local state and a behaviour that is deeper down in the tree ends up rejecting the connection which essentially discards the data you just moved into the handler.

To avoid this, make sure any "connection management" behaviour like `libp2p-connection-limits` are listed first in your `Behaviour` that has `#[derive(NetworkBehaviour)]`.

Reply:
> To avoid this, make sure any "connection management" behaviour like libp2p-connection-limits are listed first in your Behaviour that has #[derive(NetworkBehaviour)].

Thanks I'll make this change internally for now.